hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java

```diff
@@ -289,18 +304,112 @@ private MetricsRegistryImpl initMetricsRegistry() {
                 throw e;
             }
         }
+        return null;
+    }
 
-        addressProviders.add(new DefaultAddressProvider(networkConfig, addressProviders.isEmpty()));
-        return addressProviders;
+    private AddressTranslator createAddressTranslator() {
+        ClientNetworkConfig networkConfig = getClientConfig().getNetworkConfig();
+        ClientAwsConfig awsConfig = networkConfig.getAwsConfig();
+        ClientCloudConfig cloudConfig = networkConfig.getCloudConfig();
+
+        List<String> addresses = networkConfig.getAddresses();
+        boolean addressListProvided = addresses.size() != 0;
+        boolean awsDiscoveryEnabled = awsConfig != null && awsConfig.isEnabled();
+        boolean discoverySpiEnabled = discoveryService != null;
+        String cloudToken = properties.getString(HAZELCAST_CLOUD_TOKEN);
+        if (cloudToken != null && cloudConfig.isEnabled()) {
+            throw new IllegalStateException("Ambiguous hazelcast.cloud configuration. "
+                    + "Both property and configuration provided together. Use only one.");
+        }
+        boolean hazelcastCloudEnabled = cloudToken != null || cloudConfig.isEnabled();
+        isDiscoveryConfigurationConsistent(addressListProvided, awsDiscoveryEnabled,
+                discoverySpiEnabled, hazelcastCloudEnabled);
+
+        if (awsDiscoveryEnabled) {
+            try {
+                return new AwsAddressTranslator(awsConfig, loggingService);
+            } catch (NoClassDefFoundError e) {
+                ILogger logger = loggingService.getLogger(HazelcastClient.class);
+                logger.warning("hazelcast-aws.jar might be missing!");
+                throw e;
+            }
+        } else if (discoverySpiEnabled) {
+            return new DiscoveryAddressTranslator(discoveryService,
+                    getProperties().getBoolean(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED));
+        } else if (hazelcastCloudEnabled) {
+            String token;
+            if (cloudConfig.isEnabled()) {
+                token = cloudConfig.getToken();
+            } else {
+                token = cloudToken;
+            }
+            return new HazelcastCloudAddressTranslator(token, getConnectionTimeoutMillis(), loggingService);
+        }
+
+        return new DefaultAddressTranslator();
+    }
+
+    @SuppressWarnings("checkstyle:booleanexpressioncomplexity")
+    private void isDiscoveryConfigurationConsistent(boolean addressListProvided, boolean awsDiscoveryEnabled,
+                                                    boolean discoverySpiEnabled, boolean hazelcastCloudEnabled) {
+        int count = 0;
+        if (addressListProvided) {
+            count++;
+        }
+        if (awsDiscoveryEnabled) {
+            count++;
+        }
+        if (discoverySpiEnabled) {
+            count++;
+        }
+        if (hazelcastCloudEnabled) {
+            count++;
+        }
+        if (count > 1) {
```

Comment:
I believe that our discussion resulted that we may allow member address list together with the discovery, right?

Reply:
It was vice verse. You can check the final decision from related Design Document